samples/subsys/llext/edk/app/src/main.c

```diff
@@ -0,0 +1,188 @@
+{
+	int (*start_fn)(void) = p1;
+
+	printk("[app]Thread %p created to run extension [%s], at %s\n",
+	       k_current_get(), (char *)p2,
+	       k_is_user_context() ? "userspace." : "privileged mode.");
+
+	start_fn();
+	printk("[app]Thread %p done\n", k_current_get());
+}
+
+void k_sys_fatal_error_handler(unsigned int reason, const z_arch_esf_t *esf)
+{
+	int i;
+
+	printk("[app]Fatal handler! Thread: %p\n", k_current_get());
+
+	for (i = 0; i < max_extension_thread_idx; i++) {
+		if (extension_threads[i].thread == k_current_get()) {
+			unload(&extension_threads[i].ext);
+		}
+	}
+}
+
+void run_extension_on_thread(void *ext_inc, size_t ext_len,
+			     struct k_mem_domain *domain,
+			     struct k_thread *thread,
+			     k_thread_stack_t *stack,
+			     struct k_heap *heap,
+			     const char *name,
+			     k_tid_t subscriber_thread_id,
+			     int flag)
+{
+	int (*start_fn)(void);
+	struct llext **ext = &extension_threads[max_extension_thread_idx].ext;
+
+	printk("[app]Loading extension [%s].\n", name);
+	start_fn = load(name, ext, ext_inc, ext_len);
+
+	llext_add_domain(*ext, domain);
+
+	k_thread_create(thread, stack, USER_STACKSIZE,
+			user_function, start_fn, (void *)name, NULL,
+			-1, flag | K_INHERIT_PERMS, K_FOREVER);
+	k_mem_domain_add_thread(domain, thread);
+	k_mem_domain_add_thread(domain, subscriber_thread_id);
+
+	k_thread_heap_assign(thread, heap);
+
+	extension_threads[max_extension_thread_idx].thread = thread;
+	max_extension_thread_idx++;
+
+	k_thread_start(thread);
+}
+
+int main(void)
+{
+	struct k_mem_domain domain1, domain2, domain3, kdomain1;
+
+#ifndef SDK_BUILD
+	k_tid_t subscriber_thread_id = start_subscriber_thread();
+#endif
+	/* This and all other similar sleeps are here to provide a chance for
+	 * the newly created thread to run.
+	 */
+	k_sleep(K_MSEC(1));
+
+	k_mem_domain_init(&domain1, 0, NULL);
+	k_mem_domain_init(&domain2, 0, NULL);
+	k_mem_domain_init(&domain3, 0, NULL);
+	k_mem_domain_init(&kdomain1, 0, NULL);
+
+#ifndef SDK_BUILD
+#ifndef CONFIG_LLEXT_EDK_USERSPACE_ONLY
+	run_extension_on_thread(kext1_inc, kext1_len, &kdomain1,
+				&kernel_thread1, kernel_stack1, &kernel_heap1,
+				"kext1", subscriber_thread_id, 0);
+	k_sleep(K_MSEC(1));
+#endif
+	run_extension_on_thread(ext1_inc, ext1_len, &domain1, &user_thread1,
+				user_stack1, &user_heap1, "ext1",
+				subscriber_thread_id, K_USER);
+	k_sleep(K_MSEC(1));
+	run_extension_on_thread(ext2_inc, ext2_len, &domain2, &user_thread2,
+				user_stack2, &user_heap2, "ext2",
+				subscriber_thread_id, K_USER);
+	k_sleep(K_MSEC(1));
+	run_extension_on_thread(ext3_inc, ext3_len, &domain3, &user_thread3,
+				user_stack3, &user_heap3, "ext3",
+				subscriber_thread_id, K_USER);
+#endif
+
+	k_sleep(K_FOREVER);
+
+	return 0;
+}
+
+/* checkpatch.pl complains about these EXPORT_SYMBOL - ideally, they should
+ * be close to the symbol, not sneakily exported.
+ */
+EXPORT_SYMBOL(z_arm_thread_is_in_user_mode);
```

Comment:
These exports probably should exist, there’s the export_syscall macro as well to make atleast those look nicer

Reply:
Had some issues with this one, if you could please look at this patch: https://github.com/zephyrproject-rtos/zephyr/pull/69831/commits/9db9a15b25fe2c0fe13f3e62b7bbadd649959349 (`syscalls: Introduce __syscall_generated`).
If willing to not follow the spirit of `checkpatch.pl` rule here, I guess that using `EXPORT_SYSCALL` would trick it, and allowing us export the symbol directly in the application.

Reply:
Actually that's pretty clever, maybe instead of having EXPORT_SYSCALL at all use the gen_syscall.py script to generate all exports for all syscalls? Is there actually a syscall we *wouldn't* want to export?

Reply:
Hmmm - my latest headache was when some subsystem that needed syscalls were not built. But maybe just EXPORT_SYSCALL would be enough to prevent syscalls being optimised away - will do a test here.
Note that some symbols which are not syscalls would need to be exported via EXPORT_SYMBOL anyway (like `z_arm_thread_is_in_user_mode`).

Reply:
Give a try at this idea at #70324